src/Control/Monad/Aff.js

```diff
@@ -293,3 +293,70 @@ exports._liftEff = function (nonCanceler, e) {
     return nonCanceler;
   };
 }
+
+exports._tailRecM = function (isLeft, f, a) {
+  return function(success, error) {
+    return function go(acc) {
+      var result, status, canceler;
+
+      // Observes synchronous effects using a flag.
+      //
+      //   status = 0 (unresolved status)
+      //   status = 1 (synchronous effect)
+      //   status = 2 (asynchronous effect)
+      //
+      // The inner Aff is run, and if the status is still 0, we have observed
+      // a synchronous effect. The result is stashed for later synchronous
```

Comment:
The order of synchronous effects are:

* Initialize status to 0
* Run Aff
* Observe status

If the status changed between initializing and observing it, our Aff has resolved synchronously. When resolving synchronously, we check that the status has NOT changed from its initial state. You have to observe the change in two places: during resolution and, after launching the Aff.